### ISSUE TITLE
Error feedback when applying LSTM on input of wrong size

### ISSUE THREAD

**Issue by krstopro** [pytorch/pytorch#5860]
Applying LSTM (not sure if this is the issue with other modules) on input of wrong size does not give proper feedback. For example, the following code

```
import torch
import torch.autograd as autograd
import torch.nn as nn

lstm = nn.LSTM(3, 2)
hidden = (autograd.Variable(torch.zeros(3, 2)), autograd.Variable(torch.zeros(3, 2)))
input = autograd.Variable(torch.randn(6, 4, 5))
lstm(input, hidden)
```

raises ```NameError: name 'fn' is not defined```. It would be better if we received an error that says "expected this, got that" as it would help us debug the code. May I kindly ask you to have a look? Thanks.

**Comment by Stonesjtu** [comment]
You'd better add your pytorch version and full traceback information.
Your code runs into a `size mismatch` error (0.3.0), by the way.

**Comment by krstopro** [comment]
Python 3.6, Pytorch version 0.3.1.
Full traceback:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 357, in __call__
    result = self.forward(*input, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 190, in forward
    self.check_forward_args(input, hx, batch_sizes)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 141, in check_forward_args
    fn.input_size, input.size(-1)))
NameError: name 'fn' is not defined
```

**Comment by Stonesjtu** [comment]
plz refer to #4339.

**Comment by krstopro** [comment]
Thanks!

**Comment by Stonesjtu** [comment]
@apaszke perhaps you should back-port this bug fix to 0.3.1.

**Comment by apaszke** [comment]
I don’t think we’ll be making any more 0.3 releases. We’re getting close to 0.4
